SPRXCLT-10: wait for socket connection before reading input stream

alexanderchan-scality opened the pull request.

alexanderchan-scality committed:
SPRXCLT-10: wait for socket connection before reading input stream

bert-e commented:
# Hello alexanderchan-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/sproxydclient/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue SPRXCLT-10 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.0.9`


Please check the `Fix Version/s` of SPRXCLT-10, or the target
branch of this pull request.

jonathan-gramain reviewed (dismissed):
LGTM, it does not solve cases where sproxyd would stop processing the request after the connection is established, but I think it should solve the main case where sproxyd does not accept the connection in the first place.

alexanderchan-scality commented:
@jonathan-gramain 
I see.  in the case that the connection fails mid process, the request should fail immediately without attempting failover; failover would not be possible without storing the input stream in memory

alexanderchan-scality commented:
@jonathan-gramain 

* added changes to handle the case PUT requests fail mid process
* added logic to only shift when current endpoint matches the failing endpoint; this is to  address the case that batch delete failover causes multiple bootstrap list shifts.
* added test for failover

rahulreddy commented on lib/sproxyd.js:
```diff
@@ -283,27 +315,44 @@ class SproxydClient {
+
+            request.on('socket', socket => {
+                socket.on('connect', () => {
+                    log.info('Connection established', {
```
This should be debug level

rahulreddy commented on lib/sproxyd.js:
```diff
@@ -327,14 +379,15 @@ class SproxydClient {
                     return callback(err);
                 }
-                log.debug('success sending sproxyd request', {
+                log.info('success sending sproxyd request', {
```
This should be debug level

rahulreddy commented on tests/unit/sproxyd.js:
```diff
@@ -282,21 +341,86 @@ describe('Sproxyd client', () => {
+    let serverWithError;
+    let server;
+
+    before('Create the server', done => {
```
Please add another test testing the failover that it succeeds. Also a stress test ensuring requests do not fail will be nice

alexanderchan-scality replied:
the success failover  tests are performed as part of this block https://github.com/scality/sproxydclient/blob/d84a5a09b85b0dee55a78056b05c96fd29730679/tests/unit/sproxyd.js#L150

jonathan-gramain commented on lib/sproxyd.js:
```diff
@@ -312,13 +361,16 @@ class SproxydClient {
                     });
                 }
             });
+            return;
```
Why the `return` here?

alexanderchan-scality replied:
forgotten to remove this in my attempt to move the else block one tab left...decided to forgo that change to reduce line changes

alexanderchan-scality replied:
will fix

jonathan-gramain commented on lib/sproxyd.js:
```diff
@@ -226,6 +233,11 @@ class SproxydClient {
         });
 
         let receivedResponse = false;
+        // keep track of the current endpoint prior to request.
```
It's not clear to me why we need this change, could you try to clarify?

alexanderchan-scality replied:
this is to handle the issue with `batchDelete` failovers in which the bootstrap list is shifted multiple times, once per parallel request. keeping track of the request endpoint will help with determining if the bootstrap list should be shifted to prevent erroneous shifts

jonathan-gramain commented on lib/sproxyd.js:
```diff
@@ -319,6 +368,8 @@ class SproxydClient {
                 : contentType;
             const request = _createRequest(req, log, (err, response) => {
                 if (err) {
+                    // non-streaming are always retryable;
```
```suggestion
                    // non-streaming are always retryable
```

alexanderchan-scality commented:
/approve

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue SPRXCLT-10 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.0.9`


Please check the `Fix Version/s` of SPRXCLT-10, or the target
branch of this pull request.




*The following options are set:* **approve**

alexanderchan-scality commented:
ping

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.0`



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.10`
* `development/7.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.0`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.10`
* `development/7.4`



Please check the status of the associated issue SPRXCLT-10.

Goodbye alexanderchan-scality.
